scripts/copyFilesUtils.mjs

```diff
@@ -116,6 +121,38 @@ export async function createPackageFile() {
     newPackageData.types = './index.d.ts';
   }
 
+  if (!skipExportsField) {
+    newPackageData.exports = {};
+
+    const hasIndexMjs = fse.existsSync(path.resolve(buildPath, './index.mjs'));
+    if (hasIndexMjs) {
+      // Asumes the types file and node build are set up correctly
+      newPackageData.exports['.'] = {
+        types: './index.d.ts',
+        import: './index.mjs',
+        default: './node/index.js',
+      };
+    }
+
+    const hasNestedIndexFiles = glob.sync('*/index.mjs', { cwd: buildPath }).length > 0;
+    if (hasNestedIndexFiles) {
+      // Asumes the types files and node build are set up correctly
+      newPackageData.exports['./*'] = {
+        types: './*/index.d.ts',
+        import: './*/index.mjs',
+        default: './node/*/index.js',
+      };
+    }
+
+    if (!hasIndexMjs && !hasNestedIndexFiles) {
+      // Ensure that the packages we expect to have an exports field are set up correctly
+      // It avoids, for example, a top level non index file breaking the configuration
```

Comment:
Is the purpose so that we can exclude the icons package? What if we just skip the exports field generation if the package.json already contains an exports field? The package.json exports field would act as an override, that way we can special-case the icons package.

Reply:
No, to exclude the icons and downloads-tracker packages, I implemented `--skipExportsField`, which we'll probably be able to delete when we move the icons package to ESM-only.

The goal for this is to avoid the following example:

- Someone inadvertently adds a non-index file to the package's `src` root.
- The `build.mjs` detects it, and [builds the package with CJS on the root](https://github.com/mui/material-ui/blob/next/scripts/build.mjs#L47-L70).
- The exports structure changes silently; we don't notice, it's released, and things break.

All packages have ESM on the top level (and CJS under `node/`) besides the icons package. To simplify the implementation, I assumed ESM on the top level and threw an error if it was not followed. This is this error.

We could accommodate ESM and CJS at the top level, but it's not worth it if none of the packages have CJS at the top level.

As I write this, I notice that another way to do this would be to always build ESM on the top level and CJS under `node/`, avoiding the example. Or maybe this safeguard is not necessary at all 😅.

Reply:
I removed the check to be able to support the `@mui/types` exports field. I don't think it was the correct implementation.

I do think we need a way to test that our exports configuration doesn't break. Maybe we already have tests that cover this which I don't know of?